Add pendulum CNT-91 frequency counter.

bleykauf opened the pull request:
I added some functionality for a new device: the pendulum CNT-91 frequency counter.

There is only a small fraction of commands implemented so far. It has a convenience function for taking time series of frequency data (that's what I use the device for). Validators and value mapping are implemented for the commands.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        """
+        # Infinite loop occures if not redefined.
+        return super().ask(command).rstrip("\n")
+
```
Why infinite loop occurs?

bleykauf replied:
Infinite loop is the wrong term. When I ran a `Procedure` without modifying `ask` it got stuck.

But I will have a closer look at this, i.e. check what is actually returned by the device and why it is not properly taken care of by the the `ask` method of  `Instrument`.

bleykauf replied:
So it's a simple line ending thing. I thought that this was automatically stripped by `Instrument.ask`. 

@LongnoseRob clarified this in https://github.com/pymeasure/pymeasure/pull/508#discussion_r742808399.

Should be clearer now.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+    def operation_complete(self):
+        """Is True if operation is complete."""
+        return self.ask("*OPC?") == "1"
+
```
SCPI instruments can use the `complete` property. I think there's no need to define this method.

bleykauf replied:
I did have some trouble with using `complete`. At least for the CNT-91 it did not return "1" but just a property object.

Will check more thoroughly and report back.

bleykauf replied:
Could not reproduce so the problem I had earlier must have been something else.

Fixed in 6804299.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        :return : Frequency values from the buffer.
+        """
+        while not self.operation_complete:
+            # Wait until the buffer is filled.
```
`self.complete` should do the job

bleykauf replied:
Fixed in 6804299.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        ),
+        validator=strict_discrete_set,
+        values={"A": "EXT1", "B": "EXT2", "E": "EXT4", "IMM": "IMM"},
+        map_values=True,
```
Just a curiosity, why mapping names to A, B and E ? Do you need to match instrument labelling?

bleykauf replied:
Yes, A, B, E are the labels on the instrument.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        ":ACQ:APER?",
+        ":ACQ:APER %f",
+        "Gate time for one measurement.",
+        validator=strict_range,
```
I think you could specify the units.

bleykauf replied:
True, fixed in e4d4d0c.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        strict_discrete_set,
+        values=["ON", "OFF"],
+        get_process=lambda v: "ON" if v else "OFF",
+    )
```
What about mapping ON and OFF to 1 and 0?

bilderbuchi replied:
or calling it `interpolator_autocalibrated` and accepting boolean True/False.

bleykauf replied:
If using boolean for this kind of on/off switches I can change that. 

With this instrument you can set this switch with either "ON", "OFF" or 1 and 0 using SCPI. I am not sure what would be the best way here: should it be possible to also pass 1 and 0 in the python implementation? What about True/False for setting?

bleykauf replied:
I think in the end accepting and returning booleans is the cleanest and most pythonic method. I changed this also for `continuous` in 0a0415b.

msmttchr reviewed (changes requested):
@bleykauf thanks for your contribution.
I made few comments and suggestions.
I think some further review will be needed.

LongnoseRob commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+    external_arming_start_slope = Instrument.control(
+        "ARM:SLOP?",
+        "ARM:SLOP %s",
+        "Set slope for the start arming condition.",
```
This should be a docstring if I am not mistaken

bilderbuchi replied:
Do you mean a triple-quote delimited string `"""bla bla"""`? If so, that is only Python's way of defining an optionally multiline string. A normal string is fine here.

LongnoseRob replied:
Yes exactly.
I had some issues with the display of short docstrings in the documentation in the past and have defaulted to use triple-qouted docstrings with a new line at the end.

bleykauf replied:
I don't have a strong opinion either way. Changed to triple quotes since in 1704e41 since it seems a little bit more of the standard formatting and number of lines is the same.

bilderbuchi commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+    continuous = Instrument.control(
+        "INIT:CONT?",
+        "INIT:CONT %s",
+        "Turn continious measurement 'ON' or 'OFF'",
```
*continuous

bleykauf replied:
I have not yet lost hope completely that at some point in the future I will be able to write this word correctly... Fixed in b158c7c.

bleykauf commented on AUTHORS.txt:
```diff
@@ -41,3 +41,4 @@ Nicola Corna
 Robert Eckelmann
 Sam Condon
 Andreas Maeder
+Bastian Leykauf
```
Yes, I have forked and branched before I was added to the author file. I still don't really get why there is a conflict, as the files look the same. Anyway, this can be resolved upon merge.

msmttchr replied:
I could suggest to visit you branch dev/pendulum-cnt91 with web browser and click on fetch upstream to realign to master changes.
![Screenshot_20211104-200306_Chrome](https://user-images.githubusercontent.com/37845722/140404054-b53ce9aa-db71-4cd2-9367-0b6297e5fe15.jpg)

bleykauf replied:
Thanks. I actually did that to my forked master branch. I thought I then integrated these changes into `bleykauf:dev/pendulum-cnt91` (where I also squashed some changes from an intermediate branch to not have too many commits with unverified code (did a lot from home without the actual device present).

I already have a new device in the pipeline, so I will try to get it right this time.

Thank you for all the feedback, it's really great contributing to this project :)

bleykauf commented on docs/api/instruments/pendulum/index.rst:
```diff
@@ -0,0 +1,12 @@
+.. module:: pymeasure.instruments.rohdeschwarz
+
```
Oops... fixed in 4e477c9 .

bleykauf commented on docs/api/instruments/pendulum/index.rst:
```diff
@@ -0,0 +1,12 @@
+########
+
+This section contains specific documentation on the Pendulum hwarz instruments that are implemented. If you are interested in an instrument not included, please consider :doc:`adding the instrument </dev/adding_instruments>`.
+
```
Another oops. Fixed in d491777.

LongnoseRob commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        """
+        Configure the counter for an array of measurements.
+
+        :param n_samples : The number of samples
+        :param channel : Measurment channel (A, B, C, E, INTREF)
```
trailing  space in  `:param n_samples :` breaks the formatting.

bleykauf replied:
Fixed in 57ef117 also in some other instances.

LongnoseRob commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,191 @@
+        :param command: command string to be sent to the instrument
+        """
+        # Infinite loop occures if not redefined.
+        return super().ask(command).rstrip("\n")
```
instead of using the `.rstrip()` here, it might be better to set `read_termination = "\r\n",` as part of the `super().__init()` call.
 more details on the terminations [here](https://pyvisa.readthedocs.io/en/latest/introduction/communication.html#getting-the-instrument-configuration-right)

bleykauf replied:
Thanks! Missed that part of the documentation. This is indeed much cleaner. Fixed in 5e2e222.

bleykauf commented:
I noticed that sometimes there is some garbage data at the end of the buffer. I added an aptional argument to check for the length of the read out buffer and trucate it if necessary, see adf50d4. Basically the entire code was written bei github Copilot (including docstring and comments) with the only input from my side was the addition of the parameter and the `if` statement at the beginning of the check...

I also removed two of the log messages that are not really useful,  653155c.

Thanks everyone for the feedback!

I think the only remaining issue is with the `AUTHORS.txt` that can be resolved upon merge.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,182 @@
+    MAX_BUFFER_SIZE = 32000  # User Manual 8-38
+
+    def __init__(self, resourceName, **kwargs):
+        super().__init__(resourceName, "Pendulum CNT-91", **kwargs)
```
```suggestion
        super().__init__(resourceName, "Pendulum CNT-91", timeout = 120000, read_termination = '\n', **kwargs)
```
I think you can apply change above and remove following two lines.

bleykauf replied:
Fixed in 3575d75.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,182 @@
+        set_process=lambda x: "ON" if x else "OFF",
+        get_process=lambda x: bool(x),
+    )
+
```
What about using
`values = {True: "1", False : "0"}` and `map_values = True` ?

bleykauf replied:
Yes, much cleaner. I thought for some reason that booleans are not allowed as dictionary keys. Also just realized that the mapping works in both directions.

Fixed in 9885993.

msmttchr commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,182 @@
+        values=[True, False],
+        set_process=lambda x: "ON" if x else "OFF",
+        get_process=lambda x: bool(x),
+    )
```
Similar comment to `continuos` property

bleykauf replied:
Fixed in 9885993.

bilderbuchi commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,184 @@
+        Read out the entire buffer.
+
+        :param expected_length: The expected length of the buffer. If more
+            data is read, values at the end are be removed. Defaults to 0,
```
```suggestion
            data is read, values at the end are removed. Defaults to 0,
```

bilderbuchi commented on pymeasure/instruments/pendulum/cnt91.py:
```diff
@@ -0,0 +1,184 @@
+            max_sample_rate = 125e3
+        else:
+            max_sample_rate = 250e3
+        sample_rate = strict_range(sample_rate, [0, max_sample_rate])
```
0 as a sample rate is really admissible? I think the lower limit is formed by the upper of `measurement_time`, i.e. 1/1000

bleykauf replied:
I agree, added a `MAX_MEASUREMENT_TIME` in 7ec480e to avoid having two dependend magic numbers in the code. Note that `Instrument.control` needs access to this value but it doesn't have access to `self`. Since I didn't find a clean build-in solution to define class attributes, I put this constant in the main body of the module.

Open for ideas of how to improve this.

bilderbuchi reviewed (changes requested):
LGTM, except for some small remarks.

bilderbuchi commented on docs/api/instruments/pendulum/index.rst:
```diff
@@ -0,0 +1,12 @@
+.. module:: pymeasure.instruments.pendulum
+
```
whoops, the above warning ^ means that you need to add a line to pymeasure/docs/api/instruments/index.rst as this is a new manufacturer [docs](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#adding-documentation)

bleykauf replied:
Fixed in 6915d29.

bilderbuchi reviewed (approved):
When this doc warning is fixed it's good to go from my side, thank you! :-)

bilderbuchi commented:
Thank you for bringing this to merge readiness! :-)
